TrafficCapture/trafficCaptureProxyServer/src/test/java/org/opensearch/migrations/trafficcapture/proxyserver/KafkaConfigurationProxyTest.java

```diff
@@ -0,0 +1,176 @@
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import eu.rekawek.toxiproxy.Proxy;
+import eu.rekawek.toxiproxy.model.ToxicDirection;
+import java.io.IOException;
+import java.net.URI;
+import java.time.Duration;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+import lombok.extern.slf4j.Slf4j;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Disabled;
+import org.junit.jupiter.api.Tag;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.function.ThrowingConsumer;
+import org.junit.jupiter.api.parallel.Execution;
+import org.junit.jupiter.api.parallel.ExecutionMode;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.EnumSource;
+import org.opensearch.migrations.testutils.SimpleHttpClientForTesting;
+
+@Slf4j
+@Tag("longTest")
+public class KafkaConfigurationProxyTest extends ContainerTestBase {
+  private static final String HTTPD_GET_EXPECTED_RESPONSE = "<html><body><h1>It works!</h1></body></html>\n";
+
+  private static final int DEFAULT_NUMBER_OF_CALLS = 3;
+
+  private static final long PROXY_EXPECTED_MAX_LATENCY_MS = Duration.ofSeconds(1).toMillis();
+
+  @ParameterizedTest
+  @EnumSource(FailureMode.class)
+  @Disabled
+  public void testCaptureProxyWithKafkaImpairedBeforeStart(FailureMode failureMode) {
+    try (var captureProxy = new CaptureProxyContainer(destinationProxyUrl, kafkaProxyUrl)) {
+      failureMode.apply(kafkaProxy);
+
+      captureProxy.start();
+
+      var latency = assertBasicCalls(captureProxy, DEFAULT_NUMBER_OF_CALLS);
+
+      assertLessThan(PROXY_EXPECTED_MAX_LATENCY_MS, latency.toMillis());
+    }
+  }
+
+  @ParameterizedTest
+  @EnumSource(FailureMode.class)
+  public void testCaptureProxyWithKafkaImpairedAfterStart(FailureMode failureMode) {
+    try (var captureProxy = new CaptureProxyContainer(destinationProxyUrl, kafkaProxyUrl)) {
+      captureProxy.start();
+
+      failureMode.apply(kafkaProxy);
+
+      var latency = assertBasicCalls(captureProxy, DEFAULT_NUMBER_OF_CALLS);
+
+      assertLessThan(PROXY_EXPECTED_MAX_LATENCY_MS, latency.toMillis());
+    }
+  }
+
+  @ParameterizedTest
+  @EnumSource(FailureMode.class)
+  @Execution(ExecutionMode.SAME_THREAD)
+  public void testCaptureProxyWithKafkaImpairedDoesNotAffectRequest_proxysRequest(
+      FailureMode failureMode) {
+    try (var captureProxy = new CaptureProxyContainer(destinationProxyUrl, kafkaProxyUrl)) {
+      captureProxy.start();
+      final int numberOfTests = 20;
+
+      // Performance is different for first few calls so throw them away
+      assertBasicCalls(captureProxy, 3);
+
+      var averageBaselineDuration = assertBasicCalls(captureProxy, numberOfTests);
+
+      failureMode.apply(kafkaProxy);
+
+      // Calculate average duration of impaired calls
+      var averageImpairedDuration = assertBasicCalls(captureProxy, numberOfTests);
+
+      long acceptableDifference = Duration.ofMillis(25).toMillis();
```

Comment:
Do you think that this a risk to become a flaky test?  How can we make it relatively stable (> 99-99.9%)?

Reply:
The observed latency is 5-10ms on all the runs I've done, that's where 25 was set to have the buffer.

Even when the machine is overloaded by concurrent tests, this is measuring the difference with vs without the proxy so it has some resiliency there. We could move it to a % of the total time if that seems more robust.

Reply:
If we see it flaky, we can add a resource lock on it to execute in isolation